File: src/main/java/cz/kamenitxan/jakon/core/model/Page.java
package cz.kamenitxan.jakon.core.model;

import cz.kamenitxan.jakon.core.function.FunctionHelper;
import cz.kamenitxan.jakon.webui.ObjectSettings;

import javax.json.Json;
import javax.json.stream.JsonGenerator;
import javax.persistence.*;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import cz.kamenitxan.jakon.webui.entity.JakonField;
import scala.collection.JavaConversions;
import scala.collection.JavaConverters;

/**
 * Created by Kamenitxan (kamenitxan@me.com) on 05.12.15.
 */
@Entity
public class Page extends JakonObject {
	@Column
	@JakonField
	private String title;
	@Column
	@JakonField
	private String content;
	@OneToOne
	@JakonField(inputTemplate = "String")
	private Page parent = null;
	@Column
	@JakonField
	private boolean showComments = false;

	public Page() {
		super(Page.class.getName());
	}

	public String getTitle() {
		return title;
	}

	public void setTitle(String title) {
		this.title = title;
	}

	public String getContent() {
		// TODO: parsovani funkci
		// (\{)((?:[a-z][a-z]+)).*?(\})

		Pattern p = Pattern.compile("(\\{)((?:[a-z][a-z]+))(.*?)(\\})");
		Matcher m = p.matcher(content);

		StringBuffer result = new StringBuffer();
		while (m.find()) {
			String funcion = m.group(1);
			String params = m.group(2);
			m.appendReplacement(result, FunctionHelper.getFunction(funcion).execute(FunctionHelper.splitParams(params)));
		}
		m.appendTail(result);
		return result.toString();
	}

	public void setContent(String content) {
		this.content = content;
	}

	public Page getParent() {
		return parent;
	}

	public void setParent(Page parent) {
		this.parent = parent;
	}

	public boolean isShowComments() {
		return showComments;
	}

	public void setShowComments(boolean showComments) {
		this.showComments = showComments;
	}

	@Override
	public String getUrl() {
		return "/page/" + title;
	}

	@Override
	public String toJson() {
		StringWriter writer = new StringWriter();
		JsonGenerator generator = Json.createGenerator(writer);

		generator.writeStartObject()
				.write(super.getId())
				.write(getTitle())
				.write(getContent())
				.write(getParent().getId())
				.writeEnd();
		generator.close();
		return writer.toString();
	}

	@Override
	public ObjectSettings objectSettings() {
		return new ObjectSettings(JavaConverters.asScalaBuffer(new ArrayList<String>()).toList(), JavaConverters.asScalaBuffer(new ArrayList<String>()).toList());
	}
}

File: src/main/java/cz/kamenitxan/jakon/core/model/Page.scala
package cz.kamenitxan.jakon.core.model

import java.io.StringWriter
import java.sql.{Connection, Statement, Types}

import cz.kamenitxan.jakon.core.template.pebble.MarkdownFilter
import cz.kamenitxan.jakon.webui.ObjectSettings
import cz.kamenitxan.jakon.webui.entity.JakonField
import javax.json.Json
import javax.persistence._

/**
  * Created by Kamenitxan (kamenitxan@me.com) on 05.12.15.
  */
class Page extends JakonObject with Ordered {
	@JakonField
	var title: String = ""
	@JakonField(inputTemplate = "textarea")
	var content: String = ""
	@ManyToOne
	@JakonField(inputTemplate = "String")
	var parent: Page = _
	@JakonField
	var showComments: Boolean = false
	@JakonField(listOrder = -96, shownInEdit = false, shownInList = false)
	override var objectOrder: Double = _
	@Transient
	@JakonField(listOrder = -96)
	override var visibleOrder: Int = _

	override def createUrl: String = "/page/" + title.replaceAll(" ", "_").toLowerCase


	override def createObject(jid: Int, conn: Connection): Int = {
		val sql = "INSERT INTO Page (id, title, parent_id, showComments, objectOrder, content) VALUES (?, ?, ?, ?, ?, ?)"
		val stmt = conn.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)
		stmt.setInt(1, jid)
		stmt.setString(2, title)
		if (parent == null) {
			stmt.setNull(3, Types.INTEGER)
		} else {
			stmt.setInt(3, parent.id)
		}
		stmt.setBoolean(4, showComments)
		stmt.setDouble(5, objectOrder)
		stmt.setString(6, content)
		executeInsert(stmt)
	}

	override def updateObject(jid: Int, conn: Connection): Unit = {
		val sql = "UPDATE Page SET title = ?, parent_id = ?, showComments = ?, objectOrder = ?, content = ? WHERE id = ?"
		val stmt = conn.prepareStatement(sql)
		stmt.setString(1, title)
		if (parent == null) {
			stmt.setNull(2, Types.INTEGER)
		} else {
			stmt.setInt(2, parent.id)
		}
		stmt.setBoolean(3, showComments)
		stmt.setDouble(4, objectOrder)
		stmt.setString(5, content)
		stmt.setInt(6, id)
		stmt.executeUpdate()
	}

	override def toJson: String = {
		val writer = new StringWriter
		val generator = Json.createGenerator(writer)
		generator.writeStartObject.write(id).write(title).write(MarkdownFilter.parseString(content)).write(parent.id).writeEnd
		generator.close()
		writer.toString
	}

	@Transient
	override val objectSettings = new ObjectSettings("fa-file-text-o")


}